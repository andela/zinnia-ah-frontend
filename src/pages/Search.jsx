import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { withRouter } from 'react-router';

//components
import { Tab, Menu, Label } from 'semantic-ui-react';
import PopularAuthorItem from '../components/presentationals/PopularAuthorItem/PopularAuthorItem';
import ArticleCard from '../components/presentationals/ArticleCard/ArticleCard';
import Loader from '../components/presentationals/Loader/Loader';
import Tag from '../components/presentationals/Tag/Tag';

//stylesheets
import './Search.scss';

//images
import { DEFAULT_USER_IMAGE_URL } from '../utils/config';

//modules
import { getTagArticles } from '../store/modules/search';

export class Search extends Component {
  tagArticleSearch = tag => {
    this.props.getTagArticles(tag);
  };
  render() {
    this.props.isLoading && <Loader text="please wait" size="large" />;
    let articles;
    if (this.props.articles.length > 0) {
      articles = this.props.articles.map((article, key) => (
        <ArticleCard key={key} article={article} />
      ));
    } else {
      articles = null;
    }

    let authors;
    if (this.props.authors.length > 0) {
      authors = this.props.authors.map(author => (
        <PopularAuthorItem
          key={author.name}
          name={`${author.firstName} ${author.lastName}` || author.username}
          url={author.username}
          image={author.image || DEFAULT_USER_IMAGE_URL}
          username={author.username}
        />
      ));
    } else {
      authors = null;
    }

    let tags;
    if (this.props.tags.length > 0) {
      tags = this.props.tags.map((tag, key) => (
        <Tag
          key={key}
          className="tag"
          value={tag.name}
          onClick={() => this.tagArticleSearch(tag.name)}
        />
      ));
    } else {
      tags = null;
    }

    const loader = this.props.isLoading && (
      <Loader size="large" text="Loading... please wait" />
    );

    const panes = [
      {
        menuItem: (
          <Menu.Item key="articles">
            Articles<Label>{this.props.articles.length}</Label>
          </Menu.Item>
        ),
        render: () => (
          <Tab.Pane>
            <div className="articles">{articles}</div>
          </Tab.Pane>
        ),
      },
      {
        menuItem: (
          <Menu.Item key="authors">
            Authors<Label>{this.props.authors.length}</Label>
          </Menu.Item>
        ),
        render: () => <Tab.Pane>{authors}</Tab.Pane>,
      },
      {
        menuItem: (
          <Menu.Item key="tags">
            Tags<Label>{this.props.tags.length}</Label>
          </Menu.Item>
        ),
        render: () => (
          <Tab.Pane>
            <div className="tags">{tags}</div>
          </Tab.Pane>
        ),
      },
    ];

    return (
      <Fragment>
        {loader}
        <div className="articles">
          <Tab panes={panes} key={panes} />
        </div>
      </Fragment>
    );
  }
}

Search.propTypes = {
  articles: PropTypes.any,
  authors: PropTypes.any,
  tags: PropTypes.any,
  isLoading: PropTypes.bool,
  users: PropTypes.any,
  location: PropTypes.object,
  getTagArticles: PropTypes.func,
};

const mapStateToProps = state => ({
  articles: state.search.successResponse.data.articles,
  authors: state.search.successResponse.data.authors,
  tags: state.search.successResponse.data.tags,
  isLoading: state.search.isLoading,
});

export default connect(
  mapStateToProps,
  { getTagArticles },
)(withRouter(Search));
