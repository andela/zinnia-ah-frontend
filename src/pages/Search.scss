@import '../assets/variables';

.search-container {
  display: flex;
  margin: 0 auto;
  padding: 0% 6%;
  width: 66.25rem;

  .form {
    display: flex;
    width: 100%;

  }

  .form-control {
    color: $primary-white;

    &:hover {
      background: $primary-white;
      color: $primary-black;
    }
  }

  Input {
    width: 80%;
  }

  .btn-dark {
    margin: 1.25rem 0 0;
    width: 20%;
  }

  .header {
    font-size: 10px;
 }
}

.articles-container,
.authors-container {
  display: flex;
  justify-content: center;

  .article,
  .author {
    border: 1px solid $primary-black;
    margin: 2rem 0;
    text-align: center;
    width: 25rem;
  }
}
