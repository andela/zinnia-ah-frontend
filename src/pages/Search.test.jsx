import React from 'react';
import { shallow } from 'enzyme';

import { Search } from './Search';

const dataProps = {
  articles: [
    {
      title: 'Messi: the GOAT',
      description: 'Description goes here',
      slug: 'Messi-the-GOAT-b75505e0-70f3-4fbd-934c-5bfc71151dd8',
      status: 'DRAFT',
      readTime: '0',
      imageThumbnail:
        'https://res.cloudinary.com/nedy123/image/upload/v1531499318/mbppkzeehprspnbbcym8.png',
      createdAt: '2019-05-23T04:57:06.905Z',
    },
    {
      title: 'How to live long',
      description: 'Description goes here',
      slug: 'How-to-live-long-c396d648-21b4-4090-bb21-22fdb8002da3',
      status: 'DRAFT',
      readTime: '0',
      imageThumbnail:
        'https://res.cloudinary.com/nedy123/image/upload/v1531499267/zfzagvwapebjvr5tzxbt.svg',
      createdAt: '2019-05-23T04:57:06.905Z',
    },
    {
      title: 'Giving back to the community',
      description: 'Description goes here',
      slug: 'Giving-back-to-the-community-d5a18ca3-4301-441e-b7b9-589027c480a5',
      status: 'DRAFT',
      readTime: '0',
      imageThumbnail:
        'https://res.cloudinary.com/nedy123/image/upload/v1531498302/w5pxscssyqtgft2lhuny.svg',
      createdAt: '2019-05-23T04:57:06.905Z',
    },
    {
      title: 'Owning your learning in life',
      description: 'Description goes here',
      slug: 'Owning-your-learning-in-life-9f852c3a-6fc0-4102-9f18-67f30fcaa2d1',
      status: 'DRAFT',
      readTime: '0',
      imageThumbnail:
        'https://res.cloudinary.com/nedy123/image/upload/v1509347819/stay_connected_iru4aq.jpg',
      createdAt: '2019-05-23T04:57:06.905Z',
    },
    {
      title: 'Highlights Test Article',
      description: 'Description goes here, Eben',
      slug: 'Hello-Article-5-842afa04-cc9e-43ad-9384-007793d3fdcb',
      status: 'DRAFT',
      readTime: '0',
      imageThumbnail:
        'https://res.cloudinary.com/nedy123/image/upload/v1531499318/mbppkzeehprspnbbcym8.png',
      createdAt: '2019-05-23T04:57:06.907Z',
    },
    {
      title: 'Magnam.',
      description:
        'Odio molestias molestiae excepturi quia quia quo. Ea aut expedita et. Assumenda et illum accusantium voluptas ipsum. Fugit earum est perferendis et natus.',
      slug: 'why-i-kill-the-bird-fde6799275c1',
      status: 'DRAFT',
      readTime: '0',
      imageThumbnail:
        'https://res.cloudinary.com/nedy123/image/upload/v1533036618/mb0zefwbaccwnsuwretb.jpg',
      createdAt: '2019-05-23T04:57:06.909Z',
    },
  ],
  authors: [
    {
      username: 'gentlejane',
      firstName: 'gentle',
      lastName: 'jane',
      image: null,
    },
    {
      username: 'igbominadeveloper',
      firstName: 'igbomina',
      lastName: 'developer',
      image:
        'https://res.cloudinary.com/emmsdan/image/upload/v1558798049/authors-haven/favour_ql6nkl.jpg',
    },
    {
      username: 'smileygiant',
      firstName: null,
      lastName: null,
      image: null,
    },
    {
      username: 'tinawhatsgood',
      firstName: 'tina',
      lastName: 'whatsgood',
      image: null,
    },
    {
      username: 'igbominadeveloperhjdhjdjhjdhjh',
      firstName: null,
      lastName: null,
      image: null,
    },
    {
      username: 'igbominadeveloperunder',
      firstName: null,
      lastName: null,
      image: null,
    },
    {
      username: 'igbominadeveloperunderhh',
      firstName: null,
      lastName: null,
      image: null,
    },
    {
      username: 'igbominadeveloperjhdj',
      firstName: null,
      lastName: null,
      image: null,
    },
    {
      username: 'igbominadeveloperhghg',
      firstName: null,
      lastName: null,
      image: null,
    },
  ],
  tags: [
    {
      name: 'technology',
    },
  ],
  tagArticleSearch: jest.fn(),
};

const emptyProps = {
  articles: [],
  authors: [],
  tags: [],
  tagArticleSearch: jest.fn(),
};
const dataComponent = shallow(<Search {...dataProps} />);
const emptyComponent = shallow(<Search {...emptyProps} />);

describe('Custom Search', () => {
  it('renders without errors when props are not empty', () => {
    expect(dataComponent).toMatchSnapshot();
  });

  it('renders without errors when props are empty', () => {
    expect(emptyComponent).toMatchSnapshot();
  });
});
