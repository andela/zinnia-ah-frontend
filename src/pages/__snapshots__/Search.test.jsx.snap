// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Custom Search renders without errors when props are empty 1`] = `
<Fragment>
  <div
    className="articles"
  >
    <Tab
      grid={
        Object {
          "paneWidth": 12,
          "tabWidth": 4,
        }
      }
      key="[object Object],[object Object],[object Object]"
      menu={
        Object {
          "attached": true,
          "tabular": true,
        }
      }
      panes={
        Array [
          Object {
            "menuItem": <MenuItem>
              Articles
              <Label>
                0
              </Label>
            </MenuItem>,
            "render": [Function],
          },
          Object {
            "menuItem": <MenuItem>
              Authors
              <Label>
                0
              </Label>
            </MenuItem>,
            "render": [Function],
          },
          Object {
            "menuItem": <MenuItem>
              Tags
              <Label>
                0
              </Label>
            </MenuItem>,
            "render": [Function],
          },
        ]
      }
      renderActiveOnly={true}
    />
  </div>
</Fragment>
`;

exports[`Custom Search renders without errors when props are not empty 1`] = `
<Fragment>
  <div
    className="articles"
  >
    <Tab
      grid={
        Object {
          "paneWidth": 12,
          "tabWidth": 4,
        }
      }
      key="[object Object],[object Object],[object Object]"
      menu={
        Object {
          "attached": true,
          "tabular": true,
        }
      }
      panes={
        Array [
          Object {
            "menuItem": <MenuItem>
              Articles
              <Label>
                6
              </Label>
            </MenuItem>,
            "render": [Function],
          },
          Object {
            "menuItem": <MenuItem>
              Authors
              <Label>
                9
              </Label>
            </MenuItem>,
            "render": [Function],
          },
          Object {
            "menuItem": <MenuItem>
              Tags
              <Label>
                1
              </Label>
            </MenuItem>,
            "render": [Function],
          },
        ]
      }
      renderActiveOnly={true}
    />
  </div>
</Fragment>
`;
